File: packages/server/src/sidecars/manager.ts
import { connect } from "net"
import type { EventBus } from "../events/bus"
import type { Logger } from "../logger"
import type { SettingsService } from "../settings/service"
import type { SideCar, SideCarKind, SideCarPrefixMode, SideCarStatus } from "../api-types"

interface SideCarManagerOptions {
  settings: SettingsService
  eventBus: EventBus
  logger: Logger
}

interface SideCarConfigRecord {
  id: string
  kind: SideCarKind
  name: string
  port: number
  insecure: boolean
  prefixMode: SideCarPrefixMode
  createdAt: string
  updatedAt: string
}

interface SideCarRuntimeRecord {
  status: SideCarStatus
}

export class SideCarManager {
  private readonly configs = new Map<string, SideCarConfigRecord>()
  private readonly runtime = new Map<string, SideCarRuntimeRecord>()

  constructor(private readonly options: SideCarManagerOptions) {
    for (const record of this.loadConfiguredSideCars()) {
      this.configs.set(record.id, record)
      this.runtime.set(record.id, { status: "stopped" })
    }

    queueMicrotask(() => {
      for (const record of this.configs.values()) {
        void this.refreshPortSideCar(record.id).catch((error) => {
          this.options.logger.warn({ sidecarId: record.id, err: error }, "Failed to probe sidecar port")
        })
      }
    })
  }

  async list(): Promise<SideCar[]> {
    await this.refreshPortStatuses()
    return Array.from(this.configs.values()).map((record) => this.toSideCar(record))
  }

  async get(id: string): Promise<SideCar | undefined> {
    if (!this.configs.has(id)) return undefined
    await this.refreshPortSideCar(id)
    return this.toSideCar(this.requireConfig(id))
  }

  async create(input: {
    kind: SideCarKind
    name: string
    port: number
    insecure: boolean
    prefixMode: SideCarPrefixMode
  }): Promise<SideCar> {
    const normalizedName = input.name.trim()
    const id = this.buildSideCarId(normalizedName)
    if (this.configs.has(id)) {
      throw new Error(`SideCar '${id}' already exists`)
    }

    const now = new Date().toISOString()
    const record: SideCarConfigRecord = {
      id,
      kind: input.kind,
      name: normalizedName,
      port: input.port,
      insecure: input.insecure,
      prefixMode: input.prefixMode,
      createdAt: now,
      updatedAt: now,
    }

    this.configs.set(record.id, record)
    this.runtime.set(record.id, { status: "stopped" })
    this.persistConfigs()
    await this.refreshPortSideCar(record.id)
    return this.toSideCar(record)
  }

  async update(
    id: string,
    input: Partial<{
      name: string
      port: number
      insecure: boolean
      prefixMode: SideCarPrefixMode
    }>,
  ): Promise<SideCar> {
    const record = this.requireConfig(id)

    record.name = typeof input.name === "string" ? input.name.trim() : record.name
    record.port = typeof input.port === "number" ? input.port : record.port
    record.insecure = typeof input.insecure === "boolean" ? input.insecure : record.insecure
    record.prefixMode = typeof input.prefixMode === "string" ? input.prefixMode : record.prefixMode
    record.updatedAt = new Date().toISOString()

    this.persistConfigs()
    await this.refreshPortSideCar(id)
    return this.toSideCar(record)
  }

  async delete(id: string): Promise<boolean> {
    const record = this.configs.get(id)
    if (!record) return false

    this.configs.delete(id)
    this.runtime.delete(id)
    this.persistConfigs()
    this.options.eventBus.publish({ type: "sidecar.removed", sidecarId: id })
    return true
  }

  async shutdown() {
    return
  }

  buildTargetOrigin(sidecar: Pick<SideCar, "port" | "insecure">): string {
    const protocol = sidecar.insecure ? "http" : "https"
    return `${protocol}://127.0.0.1:${sidecar.port}`
  }

  buildProxyBasePath(id: string): string {
    return `/sidecars/${encodeURIComponent(id)}`
  }

  buildTargetPath(id: string, incomingPath: string, search = ""): string {
    const record = this.requireConfig(id)
    const publicBase = this.buildProxyBasePath(id)
    const normalizedPath = incomingPath || publicBase

    if (record.prefixMode === "preserve") {
      return `${normalizedPath}${search}`
    }

    let stripped = normalizedPath.startsWith(publicBase) ? normalizedPath.slice(publicBase.length) : normalizedPath
    if (!stripped || stripped === "/") {
      stripped = "/"
    } else if (!stripped.startsWith("/")) {
      stripped = `/${stripped}`
    }
    return `${stripped}${search}`
  }

  private async refreshPortStatuses() {
    await Promise.all(Array.from(this.configs.values()).map((record) => this.refreshPortSideCar(record.id)))
  }

  private async refreshPortSideCar(id: string) {
    const record = this.configs.get(id)
    if (!record) return
    const isAvailable = await this.isPortAvailable(record.port)
    const current = this.runtime.get(id)
    const nextStatus: SideCarStatus = isAvailable ? "running" : "stopped"
    if (current?.status === nextStatus) {
      return
    }

    this.runtime.set(id, { status: nextStatus })
    record.updatedAt = new Date().toISOString()
    this.publish(id)
  }

  private publish(id: string) {
    const record = this.configs.get(id)
    if (!record) return
    this.options.eventBus.publish({ type: "sidecar.updated", sidecar: this.toSideCar(record) })
  }

  private toSideCar(record: SideCarConfigRecord): SideCar {
    const runtime = this.runtime.get(record.id)
    return {
      id: record.id,
      kind: record.kind,
      name: record.name,
      port: record.port,
      insecure: record.insecure,
      prefixMode: record.prefixMode,
      status: runtime?.status ?? "stopped",
      createdAt: record.createdAt,
      updatedAt: record.updatedAt,
    }
  }

  private requireConfig(id: string): SideCarConfigRecord {
    const record = this.configs.get(id)
    if (!record) {
      throw new Error("SideCar not found")
    }
    return record
  }

  private persistConfigs() {
    const sidecars = Array.from(this.configs.values()).map((record) => ({ ...record }))
    this.options.settings.mergePatchOwner("config", "server", { sidecars })
  }

  private loadConfiguredSideCars(): SideCarConfigRecord[] {
    const serverConfig = this.options.settings.getOwner("config", "server") as { sidecars?: unknown }
    const list = Array.isArray(serverConfig?.sidecars) ? serverConfig.sidecars : []
    const records: SideCarConfigRecord[] = []
    for (const item of list) {
      if (!item || typeof item !== "object") continue
      const record = item as Record<string, unknown>
      const kind = record.kind === "port" ? "port" : null
      const id = typeof record.id === "string" && record.id.trim() ? record.id.trim() : null
      const name = typeof record.name === "string" && record.name.trim() ? record.name.trim() : null
      const port = typeof record.port === "number" && Number.isInteger(record.port) ? record.port : null
      if (!kind || !id || !name || !port) continue

      const insecure = record.insecure === true
      const prefixMode = record.prefixMode === "preserve" ? "preserve" : "strip"
      const createdAt = typeof record.createdAt === "string" && record.createdAt ? record.createdAt : new Date().toISOString()
      const updatedAt = typeof record.updatedAt === "string" && record.updatedAt ? record.updatedAt : createdAt
      records.push({ id, kind, name, port, insecure, prefixMode, createdAt, updatedAt })
    }
    return records
  }

  private isPortAvailable(port: number): Promise<boolean> {
    return new Promise((resolve) => {
      const socket = connect({ port, host: "127.0.0.1" }, () => {
        socket.end()
        resolve(true)
      })
      socket.once("error", () => {
        socket.destroy()
        resolve(false)
      })
    })
  }

  private buildSideCarId(name: string): string {
    const normalized = name
      .trim()
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, "-")
      .replace(/-{2,}/g, "-")
      .replace(/^-|-$/g, "")

    if (!normalized) {
      throw new Error("SideCar name must include letters or numbers")
    }

    return normalized
  }
}

File: packages/server-rust/src/sidecars/manager.rs
use std::collections::HashMap;
use std::net::TcpStream;
use std::sync::Arc;
use std::time::Duration;
use tokio::sync::Mutex;

use crate::api_types::{SideCar, SideCarKind, SideCarPrefixMode, SideCarStatus};
use crate::events::bus::EventBus;
use crate::logger::Logger;
use crate::settings::service::{DocKind, SettingsService};

const SIDECARS_STATE_KEY: &str = "sidecars";

pub struct SideCarManager {
    configs: HashMap<String, SideCarConfigRecord>,
    runtime: HashMap<String, SideCarRuntimeRecord>,
    event_bus: EventBus,
    #[allow(dead_code)]
    logger: Logger,
    settings: Option<Arc<Mutex<SettingsService>>>,
}

#[derive(Debug, Clone)]
struct SideCarConfigRecord {
    id: String,
    kind: SideCarKind,
    name: String,
    port: u16,
    insecure: bool,
    prefix_mode: SideCarPrefixMode,
    created_at: String,
    updated_at: String,
}

#[derive(Debug, Clone)]
struct SideCarRuntimeRecord {
    status: SideCarStatus,
}

impl SideCarManager {
    pub fn new(logger: Logger, event_bus: EventBus) -> Self {
        let mut mgr = Self {
            configs: HashMap::new(),
            runtime: HashMap::new(),
            event_bus,
            logger,
            settings: None,
        };

        let ids: Vec<String> = mgr.configs.keys().cloned().collect();
        for id in ids {
            mgr.refresh_port_sidecar(&id);
        }

        mgr
    }

    pub async fn with_settings(mut self, settings: Arc<Mutex<SettingsService>>) -> Self {
        self.load_configs(&settings).await;
        self.settings = Some(settings);
        let ids: Vec<String> = self.configs.keys().cloned().collect();
        for id in ids {
            self.refresh_port_sidecar(&id);
        }
        self
    }

    pub async fn list(&mut self) -> Vec<SideCar> {
        self.refresh_port_statuses();
        self.configs.values().map(|r| self.to_sidecar(r)).collect()
    }

    pub async fn get(&mut self, id: &str) -> Option<SideCar> {
        if !self.configs.contains_key(id) {
            return None;
        }
        self.refresh_port_sidecar(id);
        self.configs.get(id).map(|r| self.to_sidecar(r))
    }

    pub async fn create(&mut self, input: SideCarCreateInput) -> Result<SideCar, String> {
        let normalized_name = input.name.trim().to_string();
        let id = self.build_sidecar_id(&normalized_name)?;
        if self.configs.contains_key(&id) {
            return Err(format!("SideCar '{}' already exists", id));
        }

        let now = chrono::Utc::now().to_rfc3339();
        let record = SideCarConfigRecord {
            id: id.clone(),
            kind: input.kind,
            name: normalized_name,
            port: input.port,
            insecure: input.insecure,
            prefix_mode: input.prefix_mode,
            created_at: now.clone(),
            updated_at: now,
        };

        self.configs.insert(record.id.clone(), record.clone());
        self.runtime.insert(record.id.clone(), SideCarRuntimeRecord { status: SideCarStatus::Stopped });
        self.persist_configs().await;
        self.refresh_port_sidecar(&record.id);
        Ok(self.to_sidecar(&record))
    }

    pub async fn update(&mut self, id: &str, input: SideCarUpdateInput) -> Result<SideCar, String> {
        let mut record = self.configs.get(id).cloned().ok_or_else(|| "SideCar not found".to_string())?;

        if let Some(name) = input.name {
            record.name = name.trim().to_string();
        }
        if let Some(port) = input.port {
            record.port = port;
        }
        if let Some(insecure) = input.insecure {
            record.insecure = insecure;
        }
        if let Some(prefix_mode) = input.prefix_mode {
            record.prefix_mode = prefix_mode;
        }
        record.updated_at = chrono::Utc::now().to_rfc3339();

        self.configs.insert(id.to_string(), record.clone());
        self.persist_configs().await;
        self.refresh_port_sidecar(id);
        Ok(self.to_sidecar(&record))
    }

    pub async fn delete(&mut self, id: &str) -> bool {
        if self.configs.remove(id).is_none() {
            return false;
        }
        self.runtime.remove(id);
        self.persist_configs().await;
        self.event_bus.publish(
            crate::api_types::WorkspaceEventPayload::SidecarRemoved {
                event_type: "sidecar.removed".to_string(),
                sidecar_id: id.to_string(),
            },
        );
        true
    }

    pub async fn shutdown(&self) {}

    pub fn build_target_origin(&self, sidecar: &SideCar) -> String {
        let protocol = if sidecar.insecure { "http" } else { "https" };
        format!("{}://127.0.0.1:{}", protocol, sidecar.port)
    }

    pub fn build_proxy_base_path(&self, id: &str) -> String {
        format!("/sidecars/{}", urlencoding(id))
    }

    pub fn build_target_path(&self, id: &str, incoming_path: &str, search: &str) -> String {
        let record = match self.configs.get(id) {
            Some(r) => r,
            None => return String::new(),
        };
        let public_base = self.build_proxy_base_path(id);
        let normalized_path = if incoming_path.is_empty() { &public_base } else { incoming_path };

        if record.prefix_mode == SideCarPrefixMode::Preserve {
            return format!("{}{}", normalized_path, search);
        }

        let stripped = if normalized_path.starts_with(&public_base) {
            normalized_path[public_base.len()..].to_string()
        } else {
            normalized_path.to_string()
        };

        let stripped = if stripped.is_empty() || stripped == "/" {
            "/".to_string()
        } else if !stripped.starts_with('/') {
            format!("/{}", stripped)
        } else {
            stripped
        };

        format!("{}{}", stripped, search)
    }

    fn refresh_port_statuses(&mut self) {
        let ids: Vec<String> = self.configs.keys().cloned().collect();
        for id in ids {
            self.refresh_port_sidecar(&id);
        }
    }

    fn refresh_port_sidecar(&mut self, id: &str) {
        let record = match self.configs.get(id) {
            Some(r) => r.clone(),
            None => return,
        };

        let is_available = is_port_available(record.port);
        let current = self.runtime.get(id);
        let next_status = if is_available { SideCarStatus::Running } else { SideCarStatus::Stopped };

        if current.map(|r| &r.status) == Some(&next_status) {
            return;
        }

        self.runtime.insert(id.to_string(), SideCarRuntimeRecord { status: next_status });
        if let Some(record) = self.configs.get_mut(id) {
            record.updated_at = chrono::Utc::now().to_rfc3339();
            self.publish(id);
        }
    }

    fn publish(&self, id: &str) {
        if let Some(record) = self.configs.get(id) {
            self.event_bus.publish(
                crate::api_types::WorkspaceEventPayload::SidecarUpdated {
                    event_type: "sidecar.updated".to_string(),
                    sidecar: self.to_sidecar(record),
                },
            );
        }
    }

    fn to_sidecar(&self, record: &SideCarConfigRecord) -> SideCar {
        let runtime = self.runtime.get(&record.id);
        SideCar {
            id: record.id.clone(),
            kind: record.kind.clone(),
            name: record.name.clone(),
            port: record.port,
            insecure: record.insecure,
            prefix_mode: record.prefix_mode.clone(),
            status: runtime.map(|r| r.status.clone()).unwrap_or(SideCarStatus::Stopped),
            created_at: record.created_at.clone(),
            updated_at: record.updated_at.clone(),
        }
    }

    async fn persist_configs(&mut self) {
        let settings = match &self.settings {
            Some(s) => s,
            None => return,
        };

        let configs_json: Vec<serde_json::Value> = self.configs.values().map(|r| {
            serde_json::json!({
                "id": r.id,
                "kind": r.kind,
                "name": r.name,
                "port": r.port,
                "insecure": r.insecure,
                "prefix_mode": r.prefix_mode,
                "created_at": r.created_at,
                "updated_at": r.updated_at,
            })
        }).collect();

        let patch = serde_json::json!({
            SIDECARS_STATE_KEY: serde_json::Value::Array(configs_json)
        });

        let mut settings_lock = settings.lock().await;
        settings_lock.merge_patch_doc(&DocKind::State, &patch);
    }

    async fn load_configs(&mut self, settings: &Arc<Mutex<SettingsService>>) {
        let mut settings_lock = settings.lock().await;
        let doc = settings_lock.get_owner(&DocKind::State, SIDECARS_STATE_KEY);
        let arr = match doc.as_array() {
            Some(a) => a,
            None => return,
        };

        for item in arr {
            let id = item.get("id").and_then(|v| v.as_str()).unwrap_or("").to_string();
            if id.is_empty() { continue; }

            let record = SideCarConfigRecord {
                id: id.clone(),
                kind: item.get("kind").and_then(|v| serde_json::from_value(v.clone()).ok()).unwrap_or(SideCarKind::Port),
                name: item.get("name").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                port: item.get("port").and_then(|v| v.as_u64()).unwrap_or(0) as u16,
                insecure: item.get("insecure").and_then(|v| v.as_bool()).unwrap_or(false),
                prefix_mode: item.get("prefix_mode").and_then(|v| serde_json::from_value(v.clone()).ok()).unwrap_or(SideCarPrefixMode::Strip),
                created_at: item.get("created_at").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                updated_at: item.get("updated_at").and_then(|v| v.as_str()).unwrap_or("").to_string(),
            };

            self.configs.insert(id, record);
        }
    }

    fn build_sidecar_id(&self, name: &str) -> Result<String, String> {
        let normalized: String = name
            .to_lowercase()
            .chars()
            .map(|c| if c.is_alphanumeric() { c } else { '-' })
            .collect::<String>()
            .chars()
            .fold(String::new(), |mut acc, c| {
                if c == '-' && acc.ends_with('-') {
                    // skip duplicate hyphens
                } else {
                    acc.push(c);
                }
                acc
            })
            .trim_matches('-')
            .to_string();

        if normalized.is_empty() {
            return Err("SideCar name must include letters or numbers".to_string());
        }

        Ok(normalized)
    }
}

#[derive(Debug)]
pub struct SideCarCreateInput {
    pub kind: SideCarKind,
    pub name: String,
    pub port: u16,
    pub insecure: bool,
    pub prefix_mode: SideCarPrefixMode,
}

#[derive(Debug)]
pub struct SideCarUpdateInput {
    pub name: Option<String>,
    pub port: Option<u16>,
    pub insecure: Option<bool>,
    pub prefix_mode: Option<SideCarPrefixMode>,
}

fn is_port_available(port: u16) -> bool {
    TcpStream::connect_timeout(
        &format!("127.0.0.1:{}", port).parse().unwrap(),
        Duration::from_millis(500),
    )
    .is_ok()
}

fn urlencoding(input: &str) -> String {
    percent_encoding::utf8_percent_encode(input, percent_encoding::NON_ALPHANUMERIC).to_string()
}
